src/cider/nrepl/middleware/util/instrument.clj

```diff
@@ -236,20 +236,38 @@
    (let [map-inner (fn [forms]
                      (map-indexed #(walk-indexed (conj coor %1) f %2)
                                   forms))
-         ;; Maps are unordered, but we can try to use the keys (and
-         ;; they're important enough that we're willing to risk
-         ;; getting the position wrong).
-         result (cond (map? form)  (into {} (map (fn [[k v]]
-                                                   [k (walk-indexed (conj coor (pr-str k)) f v)])
-                                                 form))
-                      ;; Order of sets is unpredictable, unfortunately.
-                      (set? form)  form
-                      ;; Borrowed from clojure.walk/walk
-                      (list? form) (apply list (map-inner form))
-                      (instance? clojure.lang.IMapEntry form) (vec (map-inner form))
-                      (seq? form)  (doall (map-inner form))
-                      (coll? form) (into (empty form) (map-inner form))
-                      :else form)]
+         ;; Clojure uses array-maps up to some threshold (8 currently).
+         ;; So for small maps we take advantage of that, otherwise fall
+         ;; back to the heuristic below.
+         ;; Maps are unordered, but we can try to use the keys as order
+         ;; hoping they can be compared one by one and that the user
+         ;; has specified them in that order. If that fails we fall
+         ;; back to simply instrumenting the values in the order they
+         ;; they appear in the map.
+         ;; This depends on Clojure implementation details.
+         result (cond
+                  (map? form) (if (<= (count form) 8)
+                                (into {} (map-indexed (fn [i [k v]]
+                                                        [(walk-indexed (conj coor (* 2 i)) f k)
+                                                         (walk-indexed (conj coor (inc (* 2 i))) f v)])
+                                                      form))
+                                (try
+                                  (into {} (map-indexed (fn [i [k v]]
+                                                          [(walk-indexed (conj coor (* 2 i)) f k)
+                                                           (walk-indexed (conj coor (inc (* 2 i))) f v)])
+                                                        (into (sorted-map) form)))
+                                  (catch Exception e
+                                    (into {} (map-indexed (fn [i [k v]]
+                                                            [k (walk-indexed (conj coor (inc (* 2 i))) f v)])
+                                                          form)))))
+                  ;; Order of sets is unpredictable, unfortunately.
+                  (set? form)  form
```

Comment:
Also happy to take suggestions for heuristics for set literals.

Reply:
I'm afraid nothing comes to mind.